boa/src/builtins/value/mod.rs

```diff
@@ -149,6 +151,103 @@ impl Value {
+                    Property::default().value(Self::from(length)),
+                );
+                Self::object(new_obj)
+            }
+            JSONValue::Object(obj) => {
+                let new_obj = Value::new_object(Some(&interpreter.realm.global_obj));
+                for (key, json) in obj.into_iter() {
+                    let value = Self::from_json(json, interpreter);
+                    new_obj.set_property(
+                        key,
+                        Property::default()
+                            .value(value)
+                            .writable(true)
+                            .configurable(true),
+                    );
+                }
+                new_obj
+            }
+            JSONValue::Null => Self::null(),
+        }
+    }
+
+    /// Conversts the `Value` to `JSON`.
+    pub fn to_json(&self, interpreter: &mut Interpreter) -> Result<JSONValue, Value> {
+        match *self.data() {
+            ValueData::Null => Ok(JSONValue::Null),
+            ValueData::Boolean(b) => Ok(JSONValue::Bool(b)),
+            ValueData::Object(ref obj) => {
+                if obj.borrow().kind == ObjectKind::Array {
+                    let mut arr: Vec<JSONValue> = Vec::new();
+                    obj.borrow().properties.keys().for_each(|k| {
+                        if k != "length" {
+                            let value = self.get_field(k.to_string());
+                            if value.is_undefined() || value.is_function() {
+                                arr.push(JSONValue::Null);
+                            } else {
+                                arr.push(
+                                    self.get_field(k.to_string()).to_json(interpreter).unwrap(),
+                                );
+                            }
+                        }
+                    });
+                    Ok(JSONValue::Array(arr))
+                } else {
+                    let mut new_obj = Map::new();
+                    obj.borrow().properties.keys().for_each(|k| {
+                        let key = k.clone();
+                        let value = self.get_field(k.to_string());
+                        if !value.is_undefined() && !value.is_function() {
+                            // TODO: Handle the errors
+                            new_obj.insert(key, value.to_json(interpreter).unwrap());
+                        }
+                    });
+                    Ok(JSONValue::Object(new_obj))
+                }
+            }
+            ValueData::String(ref str) => Ok(JSONValue::String(str.clone())),
+            ValueData::Rational(num) => Ok(JSONValue::Number(
+                JSONNumber::from_f64(num).expect("Could not convert to JSONNumber"),
+            )),
+            ValueData::Integer(val) => Ok(JSONValue::Number(JSONNumber::from(val))),
+            ValueData::BigInt(_) => {
+                Err(interpreter.throw_type_error("BigInt value can't be serialized in JSON")?)
```

Comment:
Is this how you would expect that the type errors be thrown?

Reply:
we would just do:
```
interpreter.throw_type_error("BigInt value can't be serialized in JSON")?;
unreachable();
```
The `throw_type_error` should always give `Err` enum field so the code bellow should be `unreachable`

Reply:
I'm not seeing how:

> The throw_type_error should always give Err enum field so the code bellow should be unreachable

is the case given that `Interpreter::throw_type_error` returns a `ResultValue` https://github.com/boa-dev/boa/blob/863abb37474d6e39ccb81e13fc56aa2739aa8d2c/boa/src/exec/exception.rs#L25-L35

Can you help me understand?

Reply:
When we construct a new `TypeError` or any `Error` type it will always give a `Err()`

Here is https://github.com/boa-dev/boa/blob/master/boa/src/builtins/error/range.rs#L43 `RangeError`

Reply:
Okay, so the compiler doesn't necessarily know that it will always be an error and that is why I wasn't getting a clean build.

Just for my understanding, how come `ResultValue` was used if the `Ok(...)` portion is never used?

Reply:
> Just for my understanding, how come `ResultValue` was used if the `Ok(...)` portion is never used?

This is just so it's nicer to throw errors, so it builtin functions we have `ResultValue` as the return type and we can do:
```rust 
fn name_of_builtin(this: &mut Value, args: &[Value], ctx: &mut Interpreter) -> ResultValue {
	ctx.throw_type_error("message")
}
```
Instead of:
```rust 
fn name_of_builtin(this: &mut Value, args: &[Value], ctx: &mut Interpreter) -> ResultValue {
	ctx.throw_type_error("message")?;
	unreachable();
}
```

Reply:
Okay, that makes sense! Thanks @HalidOdat